src/mon/Monitor.cc

```diff
@@ -466,7 +466,7 @@ int Monitor::do_admin_command(
     if (cmd_getval(cmdmap, "value", val)) {
       cmd_vec.push_back(val);
     }
-    ceph_heap_profiler_handle_command(cmd_vec, out);
+    ceph_heap_profiler_handle_command(cmd_vec, err);
```

Comment:
@kamoltat @gregsfortytwo here's the logic behind this change:

Basically, in Telemetry, we are interested in collecting heap stats from OSDs, monitors, and MDS daemons. These stats are available via the admin socket command `ceph tell <daemon>.<id> heap stats`.

For example, when running `ceph tell osd.* heap stats` in the command line, the result is this, same for monitors and mds daemons:
```
[lflores@folio01 build]$ ./bin/ceph tell osd.0 heap stats
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2022-08-10T21:08:24.907+0000 7fd6b5d07700 -1 WARNING: all dangerous and experimental features are enabled.
2022-08-10T21:08:24.917+0000 7fd6b5d07700 -1 WARNING: all dangerous and experimental features are enabled.
osd.0 tcmalloc heap stats:------------------------------------------------
MALLOC:      372576056 (  355.3 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +      3405480 (    3.2 MiB) Bytes in central cache freelist
MALLOC: +      9281792 (    8.9 MiB) Bytes in transfer cache freelist
MALLOC: +      8771872 (    8.4 MiB) Bytes in thread cache freelists
MALLOC: +      3014656 (    2.9 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    397049856 (  378.7 MiB) Actual memory used (physical + swap)
MALLOC: +       106496 (    0.1 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =    397156352 (  378.8 MiB) Virtual address space used
MALLOC:
MALLOC:           3395              Spans in use
MALLOC:             73              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
```

For OSD and MDS daemons, the string output shows in the terminal, but it is also sent back in the command reply. This output in the command reply is what the mgr can use to access this output:

out/osd.0.log in a vstart cluster after calling `ceph tell osd.0 heap stats`:
```
2022-08-10T21:06:10.931+0000 7f64394d1700 20 osd.0 97 _dispatch 0x557db760f8c0 command(tid 2: {"prefix": "heap", "heapcmd": "stats"}) v1
2022-08-10T21:06:10.931+0000 7f644f7d4700  1 -- [v2:172.21.5.151:6802/1112500,v1:172.21.5.151:6803/1112500] --> 172.21.5.151:0/4089943627 -- command_reply(tid 2: 0 osd.0 tcmalloc heap stats:------------------------------------------------
MALLOC:      372563288 (  355.3 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +      3332552 (    3.2 MiB) Bytes in central cache freelist
MALLOC: +      9226752 (    8.8 MiB) Bytes in transfer cache freelist
MALLOC: +      8879840 (    8.5 MiB) Bytes in thread cache freelists
MALLOC: +      3014656 (    2.9 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    397017088 (  378.6 MiB) Actual memory used (physical + swap)
MALLOC: +       139264 (    0.1 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =    397156352 (  378.8 MiB) Virtual address space used
MALLOC:
MALLOC:           3390              Spans in use
MALLOC:             73              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
) v1 -- 0x557dce502b00 con 0x557dc732e800
```

However, the way the monitor code is implemented, the command reply is empty after calling the admin socket command. This makes it impossible for Telemetry (via the mgr) to access this output:

out/mon.a.log after calling `ceph tell mon.a heap stats` in a vstart cluster:
```
2022-08-10T21:14:55.786+0000 7f360fac8700  1 -- [v2:172.21.5.151:40075/0,v1:172.21.5.151:40076/0] --> 172.21.5.151:0/394792304 -- command_reply(tid 2: 0 ) v1 -- 0x55dfa9dbc420 con 0x55dfa9536400
```

If we change the output stream that's passed in the monitor code from `out` to `err`, the output is visible in the command line AND in the command reply.

Reply:
We discussed this is the telemetry huddle - since this is an interface only used by devs, and the osd and mds already behave this way, I see no issue with proceeding with this PR - all the info is still available via the tell version of the command.

We can fix the 'ceph daemon' version of these commands on all daemons as a separate issue - the only time they're necessary is when the mons are down and you can't reach the daemons except by local socket, which is a rare case.

Reply:
Heap command prints with "ceph tell", but not with "ceph daemon" - https://tracker.ceph.com/issues/57119

@yaarith ^

Reply:
This could actually be a problem, if we need to investigate memory issues which are prevent monitors from forming a quorum.
We should figure out how to get the output everywhere we need it. I'm not really following the details on why that's an issue.

Reply:
@gregsfortytwo in the case you are describing, would the mons be down? As Josh explained, the only time heap stats wouldn't be available for mons via "ceph daemon" after this PR would be when they are down and you can't reach the daemons except by local socket. But this is a rare case; in all other cases, heap output would be available via "ceph tell".

But I agree that the information should be available anywhere we need it; that's why https://tracker.ceph.com/issues/57119 was created, which tracks the lack of "ceph daemon" problem in mons (after this PR), OSDs and MDS daemons (this is their current behavior). But to my understanding, this is a rare case which can be fixed separately.

Reply:
"ceph tell" only works when there's an active quorum (and the monitor is in that quorum!). Memory issues are likely to lead to a monitor not being in a quorum.

If other daemons really work the same way on these interfaces, maybe it's not a problem. But I'm pretty perplexed by the parts I'm understanding.